fe_utils/solvers/mastery.py

```diff
@@ -2,9 +2,133 @@
 """
+from fe_utils import *
+import numpy as np
+from numpy import cos,sin, pi
+import scipy.sparse as sp
+import scipy.sparse.linalg as splinalg
 from argparse import ArgumentParser
 
 
+
+def assemble(fs1,fs2, f):
+    """Assemble the finite element system for the Helmholtz problem given
+    the function space in which to solve and the right hand side
+    function."""
+
+    # Create an appropriate (complete) quadrature rule.
+
+    # Tabulate the basis functions and their gradients at the quadrature points.
+
+    # Create the left hand side matrix and right hand side vector.
+    # This creates a sparse matrix because creating a dense one may
+    # well run your machine out of memory!
+
+    A = sp.lil_matrix((fs1.node_count, fs1.node_count))
+    B = sp.lil_matrix((fs2.node_count, fs1.node_count))
+    rhs = np.zeros(fs1.node_count + fs2.node_count)
+    l = rhs[:fs1.node_count]
+
+    # Now loop over all the cells and assemble A and l
+    fe1 = fs1.element
+    fe2 = fs2.element
+    d = fs1.element.cell.dim
+    mesh = fs1.mesh
+
+    # Create a quadrature rule which is exact for (f1-f2)**2.
+    Q = gauss_quadrature(fe1.cell, 2*max(fe1.degree, fe2.degree))
+    # Q2 = gauss_quadrature(fe2.cell, 2*max(fe1.degree, fe2.degree))
+
+    # Evaluate the local basis functions at the quadrature points.
+    phi1 = fe1.tabulate(Q.points)
+    phi1_grad = fe1.tabulate(Q.points, grad=True)
+    phi2 = fe2.tabulate(Q.points)
+    phi2_grad = fe2.tabulate(Q.points, grad=True)
+
+    val = 0.
+    for c in range(mesh.entity_counts[-1]):
+        # Find the appropriate global node numbers for this cell.
+        nodes1 = fs1.cell_nodes[c, :]
+        nodes2 = fs2.cell_nodes[c, :]
+
+        # Compute the change of coordinates.
+        J = mesh.jacobian(c)
+        detJ = np.abs(np.linalg.det(J))
+        # b = basis, d = dimension, q= quadrature 
+        f_quadrature = np.einsum("b, bdq -> dq", f.values[nodes1], phi1.T)
+        l[nodes1] += np.einsum("bdq, dq -> b", phi1.T, f_quadrature * Q.weights) * detJ 
+
+        # # i = gradient component
+        # # ji dimensions of jacobian
+        # J_phi = np.einsum("ji,bdqi -> bdqi", np.linalg.inv(J.T), phi1_grad)
+        # div = np.sum(np.array([J_phi[:, j, :, j] for j in range(d)]), axis=0)
+        # # c = basis of other space
+        # B[np.ix_(nodes2, nodes1)] += np.einsum("q, qc, qb -> cb", Q.weights, phi2, div) * detJ 
+
+        div = np.array([phi1_grad[:, j, :, j] for j in range(d)])
+        Jdiv = np.einsum("ij,jkm -> km", np.linalg.inv(J.T), div)
+        B[np.ix_(nodes2, nodes1)] += np.einsum("i, ij, ik -> jk", Q.weights, phi2, Jdiv) * detJ 
+
+        # # phi_term = np.einsum("iq, jq -> ijq", phi1.T, phi1.T)
+        
+        # half_grad_term = np.einsum("li, ijkm -> ljkm", np.linalg.inv(J.T), phi1_grad.T)
+        half_grad_term = np.einsum("ji,bdqi -> bdqi", np.linalg.inv(J.T), phi1_grad)
+        half_grad_termT = np.einsum("bdqi -> biqd", half_grad_term)
```

Comment:
This also somehow looks like the wrong transpose, but I find your index numbering hard to follow. If i is the basis function index as in the notes, then this is definitely the wrong transpose.

Reply:
i is numbering the derivatives. b is the basis numbering, d is the dimension numbers/components (2), q is the quadrature numbering. I think this is right - it swaps the derivative axis with the components axis.

Reply:
You're trying to kill me, right? https://finite-element.github.io/5_functions.html#integration
i, j, k... are conventionally basis function indices. q for quadrature, and a, b, g, e (standing in for greek letters) for geometric dimensions.
But yes, swapping derivatives and components is right.

Reply:
at least its somewhat consistent! before i was randomly using ijklm for any axis i wanted. will consider changing it to help your marking!